### PR TITLE
fix: number-field improved value handling and event management

### PR DESCRIPTION
**Pull request by robarbms**
# Pull Request

## 📖 Description

Centralizes logic to validate the value being updated. Ensures that it's a number and that it's between the min and max values. Sets the proxy value and fires "input" and "change" events for the component.

Ensures that the value is within range if the min or max attributes change.

When the number-field has the readonly or disabled attribute, the step up and step down controls still work. This piggy-backs off of the hideStep attribute to remove the controls when the readonly or disabled attribute is enabled.

### 🎫 Issues

https://github.com/microsoft/fast/issues/4776

### General

- [x] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [ ] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [x] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

File: packages/web-components/fast-foundation/src/number-field/number-field.ts [hunk 1]
```diff
@@ -169,38 +171,38 @@ export class NumberField extends FormAssociatedNumberField {
...
+                value = this.max;
+            }
+
+            value = parseFloat(value.toPrecision(12)).toString();
         }
 
-        return parseFloat(value.toPrecision(12)).toString();
+        if (value != this.value) {
+            this.value = value;
+            if (this.proxy instanceof HTMLElement) {
```
Comment (scomea) [inline packages/web-components/fast-foundation/src/number-field/number-field.ts:200]: Shouldn't this be checking for HTMLInputElement as that has a "value" prop and HTMLElement does not?

File: packages/web-components/fast-foundation/src/number-field/number-field.ts [hunk 2]
```diff
@@ -169,38 +171,38 @@ export class NumberField extends FormAssociatedNumberField {
...
+                this.proxy.value = this.value;
+            }
+            this.$emit("input");
+            this.$emit("change");
```
Comment (scomea) [inline packages/web-components/fast-foundation/src/number-field/number-field.ts:204]: What does the "input" event mean?  Beforehand it was only emitted when using the step methods, now on every change?

Comment (robarbms) [reply packages/web-components/fast-foundation/src/number-field/number-field.ts:204]: HTML number fields fire both the input and change events when ever the user interacts with the field or the step controls. I'm just trying to make both events available per an issue logged. I moved the value updates, validation and event firing to ensure consistency between the different methods to update.

File: packages/web-components/fast-foundation/src/number-field/number-field.ts (hunk 1 above)
Comment (robarbms) [reply packages/web-components/fast-foundation/src/number-field/number-field.ts]: It should be HTMLInputElement. Updated

**Review by scomea — APPROVED** [review]

**Review by EisenbergEffect — APPROVED** [review]

**Pull request merged**